src/gui/settingsDialog/SettingsDialog.cpp

```diff
@@ -165,37 +167,121 @@ void SettingsDialog::initGui()
-    mTreeWidget->addTopLevelItem(hotkeys);
-    mTreeWidget->addTopLevelItem(actions);
-    mTreeWidget->addTopLevelItem(plugins);
+	mTreeWidget->addTopLevelItem(hotkeys);
+	mTreeWidget->addTopLevelItem(actions);
+	mTreeWidget->addTopLevelItem(plugins);
 	mTreeWidget->setHeaderHidden(true);
-    mTreeWidget->setItemSelected(mNavigatorItems[0], true);
-    mTreeWidget->setFixedWidth(mTreeWidget->minimumSizeHint().width() + mScaledSizeProvider->scaledWidth(100));
-    mTreeWidget->expandAll();
+	mTreeWidget->setItemSelected(mNavigatorItems[0], true);
+	mTreeWidget->setFixedWidth(mTreeWidget->minimumSizeHint().width() + mScaledSizeProvider->scaledWidth(100));
+	mTreeWidget->expandAll();
 
-    auto listAndStackLayout = new QHBoxLayout;
-    listAndStackLayout->addWidget(mTreeWidget);
-    listAndStackLayout->addLayout(mStackedLayout);
+	mSearchSettingsLineEdit->setPlaceholderText(tr("Search Settings..."));
+	mSearchSettingsLineEdit->setFixedWidth(mTreeWidget->width());
 
-    auto mainLayout = new QVBoxLayout();
-    mainLayout->addLayout(listAndStackLayout);
-    mainLayout->addLayout(buttonLayout);
+	auto settingsNavigationLayout = new QVBoxLayout();
+	settingsNavigationLayout->addWidget(mSearchSettingsLineEdit);
+	settingsNavigationLayout->addWidget(mTreeWidget);
 
-    setLayout(mainLayout);
+	auto listAndStackLayout = new QHBoxLayout;
+	listAndStackLayout->addLayout(settingsNavigationLayout);
+	listAndStackLayout->addLayout(mStackedLayout);
+
+	auto mainLayout = new QVBoxLayout();
+	mainLayout->addLayout(listAndStackLayout);
+	mainLayout->addLayout(buttonLayout);
+
+	setLayout(mainLayout);
 }
 
 void SettingsDialog::switchTab()
 {
-    mStackedLayout->setCurrentIndex(mNavigatorItems.indexOf(mTreeWidget->currentItem()));
+	mStackedLayout->setCurrentIndex(mNavigatorItems.indexOf(mTreeWidget->currentItem()));
+}
+
+void SettingsDialog::filterSettings(const QString &filterString)
+{
+	if (filterString.isEmpty()) {
+		foreach (auto navigatorItem, mNavigatorItems) {
+			navigatorItem->setHidden(false);
+		}
+		return;
+	}
+
+	for (int index = 0; index < mTreeWidget->topLevelItemCount(); ++index) {
+		filterNavigatorItem(mTreeWidget->topLevelItem(index), filterString);
+	}
+
+	for (int index = 0; index < mNavigatorItems.size(); ++index) {
+		if (!mNavigatorItems[index]->isHidden()) {
+			mTreeWidget->setCurrentItem(mNavigatorItems[index]);
+			return;
+		}
+	}
+}
+
+bool SettingsDialog::filterNavigatorItem(QTreeWidgetItem *navigatorItem, const QString &filterString)
+{
+	bool filtered{true};
+
+	if (navigatorItem->text(0).contains(filterString, Qt::CaseInsensitive)) {
+		navigatorItem->setDisabled(false);
+		for (int index = 0; index < navigatorItem->childCount(); ++index) {
+			filterNavigatorItem(navigatorItem->child(index), filterString);
+		}
+		filtered = false;
+	} else {
+		filtered = !settingsPageContainsFilterString(mStackedLayout->itemAt(mNavigatorItems.indexOf(navigatorItem))->widget(), filterString);
```

Comment:
A lot of complexity in a single line, try to extract variables, I think that would improve readability.

Reply:
Moving this to a new component and writing tests.